### DISCUSSION TITLE
Nulls vs "zero" values

### DISCUSSION THREAD

**Discussion by miguelcobain** [pocketbase/pocketbase#1626] [Q&A]
I just learned that Pocketbase casts nulls to the corresponding zero value of the respective column type. e.g, in a number column, saving a `null` results in saving a `0`.

At first sight, this seems like a really bad decision to me. So I'm asking you to enlighten me. I'm sure I'm missing the rationale of this behavior.

Imagine we have a table of `students`, and each student has a `score` number column. Some students that took the test have a score, but others didn't take the test yet, so they can't have a score.
How do I distinguish between a student that failed every question and so got a `0` score and one that simply didn't take the test yet?

I always felt like the ability to distinguish between the absence or presence of information was a critical feature in database systems.

I've found other discussions that inform when this change was introduced and what it does, but I couldn't find the reason behind this decision and ways to solve problems like the one I wrote above.

Thanks in advance.

**Comment by ganigeorgiev** [comment]
PocketBase doesn't store `null` as field value (_with exception of the `json` field type_).
You can find all available fields with the supported values in https://pocketbase.io/docs/collections/#overview.

> At first sight, this seems like a really bad decision to me. So I'm asking you to enlighten me. I'm sure I'm missing the rationale of this behavior.

It is easier to work with when dealing with client side code since you know that the field has only 1 type and it is more convenient when integrating with languages that are non-nullable by default (eg. Dart). Depending on the demand we can introduce back `null` as option but for now it is not planned.


> How do I distinguish between a student that failed every question and so got a 0 score and one that simply didn't take the test yet?

There are several workarounds that you can apply. The easiest probably would be to add a new field that indicates only whether the student has taken the test or not (or just don't create a record for your student; this way the relation "student-tests" will not exist and you'll know that the user didn't take the test). Another option could be to use `-1` for this special case.

**Reply by miguelcobain** [reply]
Thank you for your response!

> It is easier to work with when dealing with client side code since you know that the field has only 1 type and it is more convenient when integrating with languages that are non-nullable by default (eg. Dart). Depending on the demand we can introduce back null as option but for now it is not planned.

I would argue that this implementation detail should be dealt with at the Dart client library level. Would that be possible?
Furthermore, at the end of the day, even Dart users will need to represent the absence of information somehow.

> There are several workarounds that you can apply. The easiest probably would be to add a new field that indicates only whether the student has taken the test or not (or just don't create a record for your student; this way the relation "student-tests" will not exist and you'll know that the user didn't take the test).

I think, while both workarounds are definitely possible, it also increases the query/app complexity with additional columns.

> Another option could be to use -1 for this special case.

One can easily imagine a scenario where negative numbers are valid values.

> Depending on the demand we can introduce back null as option but for now it is not planned.

Is there any official place where I can voice my desire to have nulls, or is a discussion like this enough?

In summary, I really think Pocketbase should support `null` values, as it is a fundamental concept of databases in general (and not just a different way of saying "zero").

**Reply by ganigeorgiev** [reply]
> I would argue that this implementation detail should be dealt with at the Dart client library level. Would that be possible?

Is not only for Dart, it was just an example and I don't think this can be handled at client-side level because once `null` is considered a value we can no longer "generalize" it because someone could attach a business logic to the `null` value.

I don't think that having a special value or additional column increases the complexity. Actually adding support for `null` will increase the app complexity because now your api rules will have to check for both `null` and the zero value. Yes, it may not be convenient or the best option for all cases, but it is a lot easier to not have to think about the type and to know that the field can have only a single representation.

As mentioned previously, we can introduce `null` as field option but it will definitely complicate the internals, testing and documentation and at least for now I don't think it is worth the effort (_there are other critical features and blockers in the [roadmap](https://github.com/orgs/pocketbase/projects/2) that will need to be addressed first_).

**Reply by miguelcobain** [reply]
I see.

Just another comment: Ideally, I think this behaviour should be kept but under a "non-nullable" toggle in the field.

**Reply by lupus78** [reply]
I just found out about this issue. I'm baffled, why NULL as a value is not allowed. I've read the reasoning, but it makes no sense to me. As @miguelcobain 's example shows, the NULL value is very important. Having to have a designated "empty" value or a second field goes against the whole logic of SQL.
Any plans of having this introduced?

**Reply by ganigeorgiev** [reply]
@lupus78 No, and if the reasoning in https://github.com/pocketbase/pocketbase/issues/122#issuecomment-1427510028 doesn't make sense to you then I would suggest to look somewhere else.

**Comment by maurobalbi** [comment]
Another use case:
Nullable columns would allow for unique indexes ignoring null values. I'm not sure if there is an easy way to handle this now.

Edit: Found the solution #542

**Reply by ganigeorgiev** [reply]
You can do that already by specifying a `WHERE` clause to the index, for example:

```sql
CREATE UNIQUE INDEX `my_idx` ON `users` (email) WHERE email != ''
```

In any case, adding a nullable option for now remains out of the scope as I still believe that extra handling around `null` will be a major footgun and will complicate unnecessary a lot of things (as suggested in https://github.com/pocketbase/pocketbase/issues/122, in the upcoming v0.23 documentation it is highlighted that the fields use zero-defaults).
